proposals/native-lock.md

```diff
@@ -0,0 +1,121 @@
+# Native lock
+
+## Summary
+[summary]: #summary
+
+Special-case how `System.Threading.Lock` interacts with the `lock` keyword (calling its `EnterLockScope` method under the hood).
+Add static analysis warnings to prevent accidental misuse of the type where possible.
+
+## Motivation
+[motivation]: #motivation
+
+.NET 9 is introducing a new [`System.Threading.Lock` type](https://github.com/dotnet/runtime/issues/34812)
+as a better alternative to existing monitor-based locking.
+The presence of the `lock` keyword in C# might lead developers to think they can use it with this new type.
+Doing so wouldn't lock according to the semantics of this type but would instead treat it as any other object and would use monitor-based locking.
+
+## Detailed design
+[design]: #detailed-design
+
+Semantics of the lock statement ([§13.13](https://github.com/dotnet/csharpstandard/blob/9af5bdaa7af535f34fbb7923e5406e01db8489f7/standard/statements.md#1313-the-lock-statement))
+are changed to special-case the `System.Threading.Lock` type:
+
+> A `lock` statement of the form `lock (x) { ... }`
+>
+> 1. **where `x` is an expression of type `System.Threading.Lock`, is precisely equivalent to:**
+>    ```cs
+>    using (x.EnterLockScope())
+>    {
+>        ...
+>    }
+>    ```
+>    **and `System.Threading.Lock` must have the following shape:**
+>    ```cs
+>    namespace System.Threading
+>    {
+>        class Lock
+>        {
+>            public Scope EnterLockScope();
+>    
+>            public ref struct Scope
+>            {
+>                public void Dispose();
+>            }
```

Comment:
Should the spec require that `Scope` is a ref struct? Not sure if there would be much benefit in relaxing what `Scope` could be.

Reply:
Definitely possible, but it seems easier implementation-wise if we know this upfront, so we can issue `ref`-related errors, for example:

```cs
async void M()
{
    lock (_lock) { } // should be an error since `using` on a `ref struct` is disallowed in async methods
}
```

This might seem easy (just rewrite the `lock` to `using` and let it fall out automatically), but I think the implementation needs to do the lowering from `lock` to `using` later than those checks are performed, so it needs to add all those checks again for `lock` (under the condition it is used with the `Lock` type).